Fix issue #106, to make it work on a real DEC VT320

sblendorio opened the pull request:
Fix issue https://github.com/xorg62/tty-clock/issues/106

- Using A_REVERSE attribute (a reversed space), now the digit are visible also on a real hardware DEC VT320
- The blinking cursor is hidden at the start, and is re-enabled at the end of the execution. I used printf() with ANSI code because ncurses "curs_set" function did not work on the real hardware
- The small "glitches" of the date's window are now fixed throw 2 "box-drawing characters" (ACS_TTEE)
![image](https://user-images.githubusercontent.com/2118813/152182802-9172342c-5430-45f4-8b9c-9ba5a480e016.png)

sblendorio committed:
Fix issue #106

anarcat commented:
would be nice to have comments inline in the code to explain where that magic come from.

also please provide a screenshot of what it will look like in modern terminal emulators, e.g. like xterm and gnome-terminal. i know it's backwards but most people will use ttyclock there...

sblendorio committed:
Add comments and simplify some expression

sblendorio commented:
Added comments to clarify "magic strings" and simplified some expression. Tomorrow I will test it on both real VT320 and "normal" Terminal window on Xwindow screen and will post screenshot here

sblendorio commented:
Just tested it on both real DEC VT320 and terminal emulator under Xwindows after yesterday's commit.

These are the results:
![image](https://user-images.githubusercontent.com/2118813/152310252-147c8545-1831-445e-88a0-544b2ec3fb2b.png)

![image](https://user-images.githubusercontent.com/2118813/152310652-d155e72e-c07e-464f-ad9d-ca6511ec7c59.png)

![image](https://user-images.githubusercontent.com/2118813/152310310-c083e81a-c217-4229-aad6-f1b16ea4ecba.png)

![image](https://user-images.githubusercontent.com/2118813/152310387-e1490239-0dea-4dfe-8da5-bd5dc63a08c9.png)

![image](https://user-images.githubusercontent.com/2118813/152310416-04d9ba5e-6e20-4127-b99a-dbca5070ca10.png)

It seems to be O.K.

sblendorio committed:
Use pre-calculated values

anarcat commented on ttyclock.c:
```diff
@@ -218,6 +218,8 @@ void
 {
      int i, sy = y;
+     unsigned int on  = COLOR_PAIR(2) | A_REVERSE;
+     unsigned int off = COLOR_PAIR(0) | A_NORMAL;
```
those seem a little ... terse. what is "on"? color? expand the variable.

why isn't this a `#define` too?

sblendorio replied:
Ok, I'm going to use more significant names for "on/off", for example "pixel_on" and "pixel_off".
I've not used a #define because the aim is to calculate the value at the beginning of the function and use it immedialy in the inner for cycle. Since the #define is just a text replacement, it does not implement precalculated values.

sblendorio replied:
@anarcat would you use different names than "pixel_on" / "pixel_off"? Can you suggest me? For example:

- colored_reverse
- transparent_normal

would be more descriptive?

anarcat replied:
i frankly have no idea what's going on here, so i'll trust your judgement blindly.

sblendorio replied:
I will explain:

pixel_on is a bitmask made by:
- COLOR_PAIR(2), defined by "init_color" here https://github.com/xorg62/tty-clock/blob/master/ttyclock.c#L540
- A_REVERSE is the attribute that defines reverse characters (reversed space is a big "pixel")

pixel_off is anothter bitmask made by:
- COLOR_PAIR(0) which means "not visible"
- A_NORMAL is "not reverse" attribute

so I use "pixel_on" to draw the "block characters" which are reversed spaces, supported by both xterm and DECterm, and "pixel_off" to draw the "empty space".

is it good to merge it? ;)

anarcat commented:
> Just tested it on both real DEC VT320 and terminal emulator under Xwindows after yesterday's commit.
> 
> These are the results: ![image](https://user-images.githubusercontent.com/2118813/152310252-147c8545-1831-445e-88a0-544b2ec3fb2b.png)
> 
> ![image](https://user-images.githubusercontent.com/2118813/152310652-d155e72e-c07e-464f-ad9d-ca6511ec7c59.png)
> 
> ![image](https://user-images.githubusercontent.com/2118813/152310310-c083e81a-c217-4229-aad6-f1b16ea4ecba.png)
> 
> ![image](https://user-images.githubusercontent.com/2118813/152310387-e1490239-0dea-4dfe-8da5-bd5dc63a08c9.png)
> 
> ![image](https://user-images.githubusercontent.com/2118813/152310416-04d9ba5e-6e20-4127-b99a-dbca5070ca10.png)
> 
> It seems to be O.K.

awesome, thanks, looks good.

sblendorio committed:
Use more significant names for variables

anarcat commented:
On 2022-02-08 01:32:49, Francesco Sblendorio wrote:
> @sblendorio commented on this pull request.
>
>
>
>> @@ -218,6 +218,8 @@ void
>  draw_number(int n, int x, int y)
>  {
>       int i, sy = y;
> +     unsigned int on  = COLOR_PAIR(2) | A_REVERSE;
> +     unsigned int off = COLOR_PAIR(0) | A_NORMAL;
>
> I will explain:
>
> pixel_on is a bitmask made by:
> - COLOR_PAIR(2), defined by "init_color" here https://github.com/xorg62/tty-clock/blob/master/ttyclock.c#L540
> - A_REVERSE is the attribute that defines reverse characters (reversed space is a big "pixel")
>
> pixel_off is anothter bitmask made by:
> - COLOR_PAIR(0) which means "not visible"
> - A_NORMAL is "not reverse" attribute

those explanations would be tremendously more useful as comments in the
source code. ;)

> so I use "pixel_on" to draw the "block characters" which are reversed spaces, supported by both xterm and DECterm, and "pixel_off" to draw the "empty space".

then those are good names.

-- 
Perl is "some assembly required". Python is "batteries included". PHP
is "kitchen sink, but it’s from Canada and both faucets are labeled C".
                         - Alex Munroe, PHP: a fractal of bad design

sblendorio committed:
Added comment on 'pixel_on' and 'pixel_off' variables

sblendorio commented:
> 

@anarcat just added the comment for "pixel_on" and "pixel_off"

sblendorio committed:
fix typo

anarcat reviewed (approved):
awesome, LGTM.

sblendorio commented:
> awesome, LGTM.

no merge?

anarcat commented:
On 2022-02-11 01:05:41, Francesco Sblendorio wrote:
>> awesome, LGTM.
>
> no merge?

i'm actually not the owner of this repository. typically, i approve MRs,
then let them sit a while to give them a chance to see changes, then i
merge if nothing happens.
-- 
In a racist society it is not enough to be non-racist, we must be antiracist.
                        - Angela Davis

dirtslayer commented:
i ended up having to comment out the lines 70, 71
//  init_pair(0 ...
//  init_pair(1,
so that i can use gnome-terminal profiles to pick the colors.
seems like a command line option to disable colors and use terminal profile might be handy
![Screenshot_2022-04-10_19-36-18](https://user-images.githubusercontent.com/560257/162650984-df84632d-327f-470c-8fd9-fee98aedbf9c.png)

anarcat commented:
hmm... why do we mess with colors here at all then?

dirtslayer commented:
to support backward compatibility we could leave the current color options, but adding an additional command line option, perhaps --no-color or -nc could disable the color setting and use the terminals 3 colors ( background, text, highlight) whatever they are called.
